coalib/bears/Bear.py

```diff
@@ -305,6 +314,36 @@ def log_message(self, log_message, timestamp=None, **kwargs):
     def run(self, *args, dependency_results=None, **kwargs):
         raise NotImplementedError
 
+    def setup_profile(self, profiler=None):
+        filename = '{}_{}.prof'.format(self.section.name, self.name)
+        if self.profile is True:
+            profiler.dump_stats(filename)
+        elif os.path.exists(self.profile) and os.path.isdir(self.profile):
+            profiler.dump_stats(os.path.join(self.profile, filename))
+        elif not os.path.isfile(self.profile):
+            os.mkdir(os.path.join(self.profile))
+            profiler.dump_stats(os.path.join(self.profile, filename))
+        else:
+            print('directory and filename cannot be same', file=sys.stdout)
+
+    def profile_run(self, *args, profile=None, **kwargs):
+        profile = cProfile.Profile() if profile is None else profile
+        bear_results = profile.runcall(self.run, *args, **kwargs)
+        if isinstance(bear_results, collections.Iterable):
+            results = []
+            iterator = iter(bear_results)
+            while True:
+                try:
+                    result = profile.runcall(next, iterator)
+                    results.append(result)
+                except StopIteration:
+                    break
+            self.setup_profile(profiler=profile)
+            return results
+        else:
+            self.setup_profile(profiler=profile)
```

Comment:
please deduplicate the code, you use `self.setup_profile` twice in both branches.

Reply:
What it the bear return is not `iterable` I know you already said a bear which return none is a buggy bear but that bear can be profiled IMO

Reply:
that's not what I mean ;) In both branches you use `self.setup_profile`, so in any case `self.setup_profile` will be executed. You can make the code a bit more compact by moving that out of the `if` clause:

```python
if ...
   results = ... # I know this name is already used here, but choose some other good name or so.
else ...
   results = ...

self.setup_profile(profiler=profile)
return results
```